### PR TITLE
[clang][Driver][HIP] Add support for mixing AMDGCNSPIRV & concrete `offload-arch`s.

### PR DESCRIPTION
**Pull request by AlexVlx**
This removes the temporary ban on mixing AMDGCN flavoured SPIR-V and concrete targets (e.g. `gfx900`) in the same HIPAMD compilation. This is done primarily by tweaking the effective / observable triple when the target is `amdgcnspirv`, which seamlessly composes with the existing infra. The test is stolen from #75357.

### CODE REVIEW

**Review by jhuber6 — COMMENTED** [review]
So, I'm thinking that SPIR-V here fits better as a triple. I've been wanting to generalize the `--offload` option to behave more like `-fopenmp-targets`. I.e. we would have `--offload=spirv64-amd-amdhsa,amdgcn-amd-amdhsa -Xarch_device_amdgcn-amd-amdhsa --offload-arch=gfx1030,gfx90a,gfx942` for a really verbose example. I'd like a separate option that simply targets the *toolchain* and then one that creates the `BoundArch`s. If it's not set then we default to some value (whatever amdgpu-arch says?).

File: clang/lib/Driver/ToolChains/HIPAMD.cpp [hunk 1]
```diff
@@ -264,12 +264,14 @@ void HIPAMDToolChain::addClangTargetOptions(
     CC1Args.push_back("-fapply-global-visibility-to-externs");
   }
 
-  // For SPIR-V we embed the command-line into the generated binary, in order to
-  // retrieve it at JIT time and be able to do target specific compilation with
-  // options that match the user-supplied ones.
-  if (getTriple().isSPIRV() &&
-      !DriverArgs.hasArg(options::OPT_fembed_bitcode_marker))
-    CC1Args.push_back("-fembed-bitcode=marker");
+  if (getEffectiveTriple().isSPIRV()) {
```
Comment (jhuber6) [inline clang/lib/Driver/ToolChains/HIPAMD.cpp:267]: I remember seeing an existing way to do this for LTO or something but I don't remember.

**Review by AlexVlx — COMMENTED** [review]
> So, I'm thinking that SPIR-V here fits better as a triple. I've been wanting to generalize the `--offload` option to behave more like `-fopenmp-targets`. I.e. we would have `--offload=spirv64-amd-amdhsa,amdgcn-amd-amdhsa -Xarch_device_amdgcn-amd-amdhsa --offload-arch=gfx1030,gfx90a,gfx942` for a really verbose example. I'd like a separate option that simply targets the *toolchain* and then one that creates the `BoundArch`s. If it's not set then we default to some value (whatever amdgpu-arch says?). 

We don't use `offload` at the moment, that's for HIPSPV. Of course, future, fancy work is more than welcome, but this merely slots into the existing infra and current use cases.

**Comment by jhuber6** [comment]
> We don't use `offload` at the moment, that's for HIPSPV. Of course, future, fancy work is more than welcome, but this merely slots into the existing infra and current use cases.

I'm not a huge fan of smuggling what is essentially a Toolchain behind what was intended as a `-mcpu` option. I guess we already somewhat support that? I know @yaxunl was the one that made the original `--offload=` RFC / support but I don't think it ever got 100% finished. I'm just a little concerned that if we support it this way then we'll never be able to change it because the argument will be that we need it for backwards compatibility. But I guess that's already the case and this patch just simplifies it?

**Comment by AlexVlx** [comment]
> > We don't use `offload` at the moment, that's for HIPSPV. Of course, future, fancy work is more than welcome, but this merely slots into the existing infra and current use cases.
> 
> I'm not a huge fan of smuggling what is essentially a Toolchain behind what was intended as a `-mcpu` option. I guess we already somewhat support that? I know @yaxunl was the one that made the original `--offload=` RFC / support but I don't think it ever got 100% finished. I'm just a little concerned that if we support it this way then we'll never be able to change it because the argument will be that we need it for backwards compatibility. But I guess that's already the case and this patch just simplifies it?

It's how it works today, I believe: `--offload-arch` unambiguously establishes the toolchain, and we don't have a separate toolchain for AMDGCN flavoured SPIR-V (hence the rather sneaky triple based manipulation). I don't disagree with the principle, but I think that tidying this up and, possibly, allowing things like multiple triples within a single toolchain in a less sneaky way would be a pretty big lift. On the other hand, we cannot just switch to an `--offload` based flow either, because we have lots of client apps that rely on `--offload-arch` and whose build infra incorporates it.

**Comment by jhuber6** [comment]
> It's how it works today, I believe: `--offload-arch` unambiguously establishes the toolchain

It's pretty ambiguous, right now it mostly works through a combination of the file type and good guessing because the targets people care about now have distinct names. For OpenMP, you can offload to x64 CPUs or whatever and then it doesn't work via `--offload-arch` because there's not a distinct CPU name we can infer the triple based on. For HIP we just infer the triple that it's AMD, but this SPIR-V thing throws the same wrench that OpenMP has. Maybe in the future someone will want `clang++ -x cuda foo.cpp --offload-arch=gfx1030` as well, who knows.

I don't think this is a blocker, but I think it bears deep consideration on how we expect these toolchains to actually work.

**Comment by AlexVlx** [comment]
Gentle ping.

**Review by jhuber6 — APPROVED** [review]
I guess this is good for now, but we really need to work with how to actually handle multiple toolchains at once.

**Review by yxsamliu — APPROVED** [review]

**Pull request merged**